src/Microsoft.Crank.Agent/Controllers/JobsController.cs

```diff
@@ -454,20 +454,31 @@ public async Task<IActionResult> UploadBuildFile(int id)
             return Ok();
         }
 
-        private async Task SaveBodyAsync(string filename)
+        private async Task SaveBodyAsync(Job job, string filename)
         {
             using var outputFileStream = System.IO.File.Create(filename);
 
+            Task task = null;
+
             if (Request.Headers.TryGetValue("Content-Encoding", out var encoding) && encoding.Contains("gzip"))
             {
                 Log.Info($"Received gzipped file content");
                 using var decompressor = new GZipStream(Request.Body, CompressionMode.Decompress);
-                await decompressor.CopyToAsync(outputFileStream, Request.HttpContext.RequestAborted);
+                task = decompressor.CopyToAsync(outputFileStream, Request.HttpContext.RequestAborted);
             }
             else
             {
                 Log.Info($"Received uncompressed file content");
-                await Request.Body.CopyToAsync(outputFileStream, Request.HttpContext.RequestAborted);
+                task = Request.Body.CopyToAsync(outputFileStream, Request.HttpContext.RequestAborted);
+            }
+
+            // Keeps controller's connection alive as PINGs might not be accepted
+            // by the agent while big files are uploaded.
+
+            while (!Request.HttpContext.RequestAborted.IsCancellationRequested && !task.IsCompleted)
+            {
+                await Task.Delay(500);
+                job.LastDriverCommunicationUtc = DateTime.UtcNow;
```

Comment:
This is the important part I assume?

Reply:
It is, is that wrong ;)

Reply:
Just checking my understanding

Reply:
Should we await the task if it completes in case there is an exception?